webots_ros2_epuck2/webots_ros2_epuck2/driver.py

```diff
@@ -0,0 +1,617 @@
+TOF_MIN_RANGE = 1.0
+INFRARED_MAX_RANGE = 0.04
+INFRARED_MIN_RANGE = 0.009
+GROUND_MIN_RANGE = 0.0
+GROUND_MAX_RANGE = 0.016
+DEFAULT_WHEEL_RADIUS = 0.02
+DEFAULT_WHEEL_DISTANCE = 0.05685
+NB_LIGHT_SENSORS = 8
+NB_GROUND_SENSORS = 3
+NB_RGB_LEDS = 4
+NB_BINARY_LEDS = 4
+NB_INFRARED_SENSORS = 8
+SENSOR_DIST_FROM_CENTER = 0.035
+# https://ieee-dataport.org/open-access/conversion-guide-solar-irradiance-and-lux-illuminance
+IRRADIANCE_TO_ILLUMINANCE = 120
+
+TOF_TABLE = [
+    [2.00, 2000.0],
+    [1.70, 1780.5],
+    [1.00, 1052.0],
+    [0.50, 531.9],
+    [0.20, 218.9],
+    [0.10, 111.0],
+    [0.05, 58.5],
+    [0.00, 19.8]
+]
+
+DISTANCE_TABLE = [
+    [0, 4095],
+    [0.005, 2133.33],
+    [0.01, 1465.73],
+    [0.015, 601.46],
+    [0.02, 383.84],
+    [0.03, 234.93],
+    [0.04, 158.03],
+    [0.05, 120],
+    [0.06, 104.09]
+]
+
+LIGHT_TABLE = [
+    [1, 4095],
+    [2, 0]
+]
+
+GROUND_TABLE = [
+    [0, 1000],
+    [0.016, 300]
+]
+
+DISTANCE_SENSOR_ANGLE = [
+    -15 * pi / 180,   # ps0
+    -45 * pi / 180,   # ps1
+    -90 * pi / 180,   # ps2
+    -150 * pi / 180,  # ps3
+    150 * pi / 180,   # ps4
+    90 * pi / 180,    # ps5
+    45 * pi / 180,    # ps6
+    15 * pi / 180,    # ps7
+]
+
+
+def euler_to_quaternion(roll, pitch, yaw):
+    """Source: https://computergraphics.stackexchange.com/a/8229."""
+    q = Quaternion()
+    q.x = sin(roll/2) * cos(pitch/2) * cos(yaw/2) - \
+        cos(roll/2) * sin(pitch/2) * sin(yaw/2)
+    q.y = cos(roll/2) * sin(pitch/2) * cos(yaw/2) + \
+        sin(roll/2) * cos(pitch/2) * sin(yaw/2)
+    q.z = cos(roll/2) * cos(pitch/2) * sin(yaw/2) - \
+        sin(roll/2) * sin(pitch/2) * cos(yaw/2)
+    q.w = cos(roll/2) * cos(pitch/2) * cos(yaw/2) + \
+        sin(roll/2) * sin(pitch/2) * sin(yaw/2)
+    return q
+
+
+def interpolate_function(value, start_x, start_y, end_x, end_y):
+    slope = (end_y - start_y) / (end_x - start_x)
+    return slope * (value - start_x) + start_y
+
+
+def interpolate_table(value, table):
+    for i in range(len(table) - 1):
+        if (value < table[i][1] and value >= table[i + 1][1]) or \
+                (value > table[i][1] and value <= table[i + 1][1]):
+            return interpolate_function(
+                value,
+                table[i][1],
+                table[i][0],
+                table[i + 1][1],
+                table[i + 1][0]
+            )
+    # Edge case, search outside of two points.
+    # This code assumes that the table is sorted in descending order
+    if value > table[0][1]:
+        # Interpolate as first
+        return interpolate_function(
+            value,
+            table[0][1],
+            table[0][0],
+            table[1][1],
+            table[1][0]
+        )
+    else:
+        # Interpolate as last
+        return interpolate_function(
+            value,
+            table[len(table) - 2][1],
+            table[len(table) - 2][0],
+            table[len(table) - 1][1],
+            table[len(table) - 1][0]
+        )
```

Comment:
This function too might be moved in a ``math_utils`` file in this package to make it available to all the packages: https://github.com/cyberbotics/webots_ros2/tree/master/webots_ros2_core/webots_ros2_core

Reply:
I also moved `get_ros_stamp()` to `utils.py`, please check if it is a proper place for it

Reply:
Yes, that's indeed good to move it there.